Fail if reading from closed KeyStoreWrapper

tvernum opened the pull request:
In #28255 the implementation of the elasticsearch.keystore was changed
to no longer be built on top of a PKCS#12 keystore. A side effect of
that change was that calling getString or getFile on a closed
KeyStoreWrapper ceased to throw an exception, and would instead return
a value consisting of all 0 bytes.

This change restores the previous behaviour as closely as possible.
It is possible to retrieve the _keys_ from a closed keystore, but any
attempt to get or set the entries will throw an IllegalStateException.

The only divergence from the previous behaviour is that "isLoaded"
will now return false if the keystore is closed, in keeping with the
"loaded and retrievable" description in the parent javadoc.

elasticmachine commented:
Pinging @elastic/es-core-infra

tvernum commented:
I've (re-)labelled this `>non-issue` on the assumption that we'll fix it before 6.3.0 is released (in which case we fix the bug in the same release that it was introduced.

If for some reason we decide not to put this into 6.3.0 then we can re-label and I'll update the release-notes accordingly.

tvernum commented:
pinging @jkakavas since you were involved in the design of keystore v3 and might be interested

rjernst commented:
I agree @jaymode. At first I thought we might want something that would allow parallel reading, but these are only read during startup, and there aren't that many. Let's start with synchronized and revisit later if necessary.

tvernum commented:
I've made three changes to this PR:

1. Added `synchronized` across the `getXXX`,`setXXX` and `close` methods.   
`setString` (etc) were problematic because the race condition would make it possible to store secret data in a keystore that was already closed, and therefore that new data would not clear zero'd out.  

2. Added `ensureOpen` and `sychronized` to  the `save` method so that it's not possible to store data that has been (or is being) zero'd out.

3. Removed the `closed == false` check from `isLoaded`

The last one is a bit of an issue. At the moment there is no method that allows a caller to check that it is safe to read from a `SecretSettings` object. The docs for `isLoaded` imply that it could be the method to do that, but `Settings.Builder` uses `isLoaded` to [check that SecretSettings have been loaded](https://github.com/elastic/elasticsearch/blob/v6.2.4/server/src/main/java/org/elasticsearch/common/settings/Settings.java#L840) before being added to the `Settings` object. We want this check to exist for safety, but it should continue to allow closed settings.

I think we may need a new `isOpen` (or similar) method, but this isn't the PR to do that.

tvernum commented:
@elasticmachine kindly test this please

tvernum commented:
run gradle build tests

tvernum commented:
run gradle build tests (again) 😿

tvernum commented:
@rjernst I've updated this since you reviewed - are you good with the changes?
